storage/src/nodestore/hash.rs

```diff
@@ -33,99 +79,95 @@ where
-                trace!("hashed {hashed:?} unhashed {unhashed:?}");
-                if hashed.len() == 1 {
-                    // we were left with one hashed node that must be rehashed
-                    let (invalidated_node_idx, (invalidated_node, invalidated_hash)) =
-                        hashed.first_mut().expect("hashed is not empty");
-                    // Extract the address from the MaybePersistedNode
-                    let addr: crate::LinearAddress = invalidated_node
-                        .as_linear_address()
-                        .expect("hashed node should be persisted");
-                    let mut hashable_node = self.read_node(addr)?.deref().clone();
-                    let original_length = path_prefix.len();
-                    path_prefix.0.extend(b.partial_path.0.iter().copied());
-                    if unhashed.is_empty() {
-                        hashable_node.update_partial_path(Path::from_nibbles_iterator(
-                            std::iter::once(*invalidated_node_idx as u8)
-                                .chain(hashable_node.partial_path().0.iter().copied()),
-                        ));
-                    } else {
-                        path_prefix.0.push(*invalidated_node_idx as u8);
-                    }
-                    let hash = hash_node(&hashable_node, path_prefix);
-                    path_prefix.0.truncate(original_length);
-                    **invalidated_hash = hash;
-                }
-                // handle the single-child case for an account special below
-                if hashed.is_empty() && unhashed.len() == 1 {
-                    Some(unhashed.last().expect("only one").0 as u8)
-                } else {
-                    None
+                } = self.ethhash_classify_children(&mut b.children)?;
+                trace!("hashed {hashed:?} unhashed {num_unhashed:?}");
+                #[expect(
+                    clippy::arithmetic_side_effects,
+                    reason = "hashed and unhashed can have at most 16 elements"
+                )]
+                let num_children = hashed.len() + num_unhashed;
+                // If there was only one child in the current account branch when previously hashed, we need to rehash it
+                if let [(child_idx, child_node, child_hash)] = &mut hashed[..] {
+                    let hash = {
+                        let mut path_guard = PathGuard::new(&mut path_prefix);
+                        path_guard.0.extend(b.partial_path.0.iter().copied());
+                        path_guard.0.push(*child_idx as u8);
+                        Self::compute_node_ethhash(child_node, &mut path_guard, num_children)
+                    };
+                    **child_hash = hash;
                 }
+
+                num_children
             } else {
-                // not a single child
-                None
+                // not an account branch - does not matter what we return here
```

Comment:
🤔

I bet we can structure this code so this statement is not reachable, or let's wrap this in an Option for readability.